### PR TITLE
Minchingtonak/threading fix

### PR DESCRIPTION
**Pull request by minchingtonak**
What would be better, the current code in fillTalonsWithProfile or the commented out code?

### CODE REVIEW

File: src/org/hammerhead226/sharkmacro/motionprofiles/Profile.java [hunk 1]
```diff
@@ -112,10 +107,8 @@ public Profile(String[][] leftProfile, String[][] rightProfile, TalonSRX leftTal
 	 *            the PID gains profile to use to execute the right motion profile
 	 */
 	public void execute(int leftGainsProfile, int rightGainsProfile) {
-		left = new ProfileHandler(leftProfile, leftTalon, leftGainsProfile);
-		right = new ProfileHandler(rightProfile, rightTalon, rightGainsProfile);
-		left.execute();
-		right.execute();
+		handler = new ProfileHandler(leftProfile, rightProfile, leftTalon, rightTalon, leftGainsProfile, rightGainsProfile);
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/Profile.java]: You could potentially do an array of profiles, talons, and gains profiles here. Not a stopper.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 2]
```diff
@@ -74,30 +96,46 @@
...
+	 * @param leftGainsProfile
+	 *            the PID gains profile to use on the left talon
+	 * @param rightGainsProfile
+	 *            the PID gains profile to use on the right talon
 	 */
-	public ProfileHandler(final double[][] profile, TalonSRX talon, int gainsProfile) {
-		this.profile = profile;
-		this.talon = talon;
-		this.gainsProfile = gainsProfile;
+	public ProfileHandler(final double[][] leftProfile, final double[][] rightProfile, TalonSRX leftTalon,
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: See above.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 3]
```diff
@@ -74,30 +96,46 @@
...
+		this.rightTalon = rightTalon;
+		this.leftGainsProfile = leftGainsProfile;
+		this.rightGainsProfile = rightGainsProfile;
 		this.executionState = ExecutionState.WAITING;
 
 		bufferThread = new Notifier(new PeriodicBufferProcessor());
 		bufferThread.startPeriodic(Constants.DT_SECONDS / 2.0);
-		this.talon.changeMotionControlFramePeriod((int) Math.round((Constants.DT_MS / 2.0)));
+		this.leftTalon.changeMotionControlFramePeriod((int) Math.round((Constants.DT_MS / 2.0)));
+		this.rightTalon.changeMotionControlFramePeriod((int) Math.round((Constants.DT_MS / 2.0)));
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: Might make sense to make that period a class constant.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 4]
```diff
@@ -20,27 +20,49 @@
...
 	 */
-	private int gainsProfile;
+	int leftProfileIndex = 0;
 
 	/**
-	 * The Talon used to execute the motion profile.
+	 * Represents the current point being streamed from the right profile to the right talon.
 	 */
-	private TalonSRX talon;
+	int rightProfileIndex = 0;
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: Pairing profiles together only makes sense if they're the same length, I don't think you need separate indexes.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 5]
```diff
@@ -147,71 +187,157 @@ public void manage() {
...
+
+		while (leftStatus.topBufferCnt < Constants.TALON_TOP_BUFFER_MAX_COUNT
+				&& leftProfileIndex < leftProfile.length) {
+
+			point.position = leftProfile[leftProfileIndex][0];
+			point.velocity = leftProfile[leftProfileIndex][1];
+			point.headingDeg = 0;
+			point.timeDur = toTrajectoryDuration((int) leftProfile[leftProfileIndex][2]);
+			point.profileSlotSelect0 = leftGainsProfile;
+			point.profileSlotSelect1 = 0;
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: Is this zeroing PID slot 1? If so, I don't think that's necessary or desired.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 6]
```diff
@@ -147,71 +187,157 @@ public void manage() {
...
+			point.zeroPos = false;
+			if (leftProfileIndex == 0) {
+				point.zeroPos = true;
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: (Of course it's named this and not `isFirstPoint` for consistency...)

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 7]
```diff
@@ -147,71 +187,157 @@ public void manage() {
...
+//
+//			leftTalon.pushMotionProfileTrajectory(leftPoint);
+//			rightTalon.pushMotionProfileTrajectory(rightPoint);
+//
+//			leftTalon.getMotionProfileStatus(leftStatus);
+//			rightTalon.getMotionProfileStatus(rightStatus);
+//
+//			leftProfileIndex++;
+//			rightProfileIndex++;
+//		}
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: I'd get rid of this.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java (hunk 5 above)
Comment (minchingtonak) [reply src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: It's not, it's to select the first slot for cascaded PID

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java (hunk 6 above)
Comment (minchingtonak) [reply src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: 👍 👍

**Comment by minchingtonak** [comment]
Current build is producing "Talon buffer is already full" warnings when run

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 8]
```diff
@@ -74,30 +76,37 @@
...
-	 * The status of {@link #talon}.
+	 * The status of {@link #leftTalon}.
+	 */
+	private MotionProfileStatus leftStatus = new MotionProfileStatus();
+
+	/**
+	 * The status of {@link #rightTalon}.
 	 */
-	private MotionProfileStatus status = new MotionProfileStatus();
+	private MotionProfileStatus rightStatus = new MotionProfileStatus();
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: Might as well do the array treatment on these, too.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 9]
```diff
@@ -74,30 +76,37 @@
...
+		this.talons = talons;
+		this.pidSlotIdxs = pidSlotIdxs;
 		this.executionState = ExecutionState.WAITING;
 
 		bufferThread = new Notifier(new PeriodicBufferProcessor());
 		bufferThread.startPeriodic(Constants.DT_SECONDS / 2.0);
-		this.talon.changeMotionControlFramePeriod((int) Math.round((Constants.DT_MS / 2.0)));
+		
+		this.talons[0].changeMotionControlFramePeriod(Constants.MOTIONCONTROL_FRAME_PERIOD);
+		this.talons[1].changeMotionControlFramePeriod(Constants.MOTIONCONTROL_FRAME_PERIOD);
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: I would loop all Talons.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 10]
```diff
@@ -127,16 +136,18 @@ private void onFinish() {
 		bufferThread.stop();
 		executorThread.stop();
 		setMode(SetValueMotionProfile.Disable);
-		talon.clearMotionProfileTrajectories();
+		talons[0].clearMotionProfileTrajectories();
+		talons[1].clearMotionProfileTrajectories();
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: I would loop all Talons.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 11]
```diff
@@ -127,16 +136,18 @@ private void onFinish() {
...
 	 * Called periodically while the motion profile is being executed. Manages the
-	 * state of the Talon executing the motion profile.
+	 * state of the Talons executing the motion profiles.
 	 */
 	public void manage() {
-		fillTalonWithMotionProfile(gainsProfile);
-		talon.getMotionProfileStatus(status);
+		fillTalonsWithMotionProfile();
+		talons[0].getMotionProfileStatus(leftStatus);
+		talons[1].getMotionProfileStatus(rightStatus);
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: I would loop all Talons.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 12]
```diff
@@ -147,71 +158,99 @@ public void manage() {
...
+	 * Sets the state of the talons.
 	 * 
 	 * @param t
 	 *            the motion profile mode to set the Talon to
 	 */
 	private void setMode(SetValueMotionProfile mode) {
 		this.currentMode = mode;
-		talon.set(ControlMode.MotionProfile, mode.value);
+		talons[0].set(ControlMode.MotionProfile, mode.value);
+		talons[1].set(ControlMode.MotionProfile, mode.value);
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: I would loop all Talons.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 13]
```diff
@@ -147,71 +158,99 @@ public void manage() {
...
 		if (profileIndex == 0) {
-			talon.clearMotionProfileTrajectories();
-			talon.configMotionProfileTrajectoryPeriod(TrajectoryDuration.Trajectory_Duration_0ms.value, 0);
-			talon.clearMotionProfileHasUnderrun(0);
+			talons[0].clearMotionProfileTrajectories();
+			talons[0].configMotionProfileTrajectoryPeriod(TrajectoryDuration.Trajectory_Duration_0ms.value, 0);
+			talons[0].clearMotionProfileHasUnderrun(0);
+			talons[1].clearMotionProfileTrajectories();
+			talons[1].configMotionProfileTrajectoryPeriod(TrajectoryDuration.Trajectory_Duration_0ms.value, 0);
+			talons[1].clearMotionProfileHasUnderrun(0);
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: I would loop all Talons.

File: src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java [hunk 14]
```diff
@@ -260,11 +299,13 @@ public boolean isFinished() {
...
 	 * {@link com.ctre.CANTalon#processMotionProfileBuffer()
-	 * processMotionProfileBufffer()} for {@link ProfileHandler#talon}.
+	 * processMotionProfileBufffer()} for {@link ProfileHandler#leftTalon} and
+	 * {@link ProfileHandler#rightTalon}.
 	 */
 	class PeriodicBufferProcessor implements java.lang.Runnable {
 		public void run() {
-			talon.processMotionProfileBuffer();
+			talons[0].processMotionProfileBuffer();
+			talons[1].processMotionProfileBuffer();
```
Comment (emmercm) [inline src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: I would loop all Talons.

Comment (emmercm) [reply src/org/hammerhead226/sharkmacro/motionprofiles/ProfileHandler.java]: Also, this may be the source of the errors.

**Pull request merged**